### PR TITLE
fix: migrate to typescript

### PR DESCRIPTION
**Pull request by anjaliahlawat**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "name": "gauge-js-template",
   "description": "Starter template for writing JavaScript tests for Gauge",
   "scripts": {
-    "test": "./node_modules/.bin/gauge run -p specs/",
+    "test": " .node_modules/.bin/tsc && ./node_modules/.bin/gauge run -p specs/",
```
Comment (h2grover) [inline package.json]: let's create a command for this ( i.e. this is called target )
```
    "tsc": "./node_modules/.bin/tsc",
    "test": " yarn run tsc && ./node_modules/.bin/gauge run -p specs/",
```

File: package.json [hunk 2]
```diff
@@ -15,18 +15,23 @@
   "dependencies": {
     "@getgauge/cli": "^1.1.7",
     "dotenv": "^8.2.0",
+    "gauge-ts": "^0.1.0",
```
Comment (h2grover) [inline package.json]: I know we have `gauge-ts` but let's not use it.

File: src/lib/PageWrapper.ts [hunk 3]
```diff
@@ -1,28 +1,31 @@
-const { v4: uniqueId } = require("uuid");
+import * as uniqueId from "uuid";
+import pageType from "./Types";
```
Comment (h2grover) [inline src/lib/PageWrapper.ts]: puppeteer provides its types by default. please check how to use it

File: src/lib/Types.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+type pageType = {
```
Comment (h2grover) [inline src/lib/Types.ts]: use puppeteer types which are provided out of the box.

File: webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
```
Comment (h2grover) [inline webpack.config.js]: webpack is not needed in our case.

File: env/default/js.properties [hunk 6]
```diff
@@ -7,4 +7,4 @@ test_timeout = 10000
 DEBUG = false
 
 # Comma seperated list of dirs. path should be relative to project root.
-STEP_IMPL_DIR = src/tests
+STEP_IMPL_DIR = build/tests
```
Comment (h2grover) [inline env/default/js.properties:10]: yes this correct approach

**Review by h2grover — CHANGES_REQUESTED** [review]
great job on the migration but in our case, we will not use `gauge-ts` we will simply rename the file to `.ts` and refactor that file.

here is the sample ts file. 

image

**Comment by h2grover** [comment]
@anjaliahlawat you may ask why not `gauge-ts` in our client project.

1. gauge-ts is new and not that stable/released officially.
1. We were working on gauge-js and it was a difficult and redundant task to update our code as per `gauge-ts` as step implementation syntax is different.
1. So we took to the approach to rename the `.js` file to `.ts` file and refactor the code for types only.

**Comment by anjaliahlawat** [comment]
> @anjaliahlawat you may ask why not `gauge-ts` in our client project.
> 
> 1. gauge-ts is new and not that stable/released officially.
> 2. We were working on gauge-js and it was a difficult and redundant task to update our code as per `gauge-ts` as step implementation syntax is different.
> 3. So we took to the approach to rename the `.js` file to `.ts` file and refactor the code for types only.

I did wonder on that and it seemed gauge-ts is new. But I was stuck at creating step definition. I also checked gauge-ts template provided by gauge itself and there also gauge-ts was used, so eventually I used it here.

File: webpack.config.js (hunk 5 above)
Comment (anjaliahlawat) [reply webpack.config.js]: Won't we need it for configuring our loaders for typescript?

**Comment by h2grover** [comment]
@anjaliahlawat we will be using `build/test` folder. after `yarn run tsc`. what was the issue you were facing?

**Comment by anjaliahlawat** [comment]
> @anjaliahlawat we will be using `build/test` folder. after `yarn run tsc`. what was the issue you were facing?

@h2grover my specification files are not able to detect their respective step definitions.

File: package.json (hunk 1 above)
Comment (anjaliahlawat) [reply package.json]: Shouldn't it be like this -
```
"test": "tsc && ./node_modules/.bin/gauge run -p specs/"
```

Comment (h2grover) [reply package.json]: if that works 👍

**Comment by h2grover** [comment]
> > @anjaliahlawat we will be using `build/test` folder. after `yarn run tsc`. what was the issue you were facing?
> 
> @h2grover my specification files are not able to detect their respective step definitions.

can you please refactor as per suggested changes then I will cross check

**Comment by anjaliahlawat** [comment]
> great job on the migration but in our case, we will not use `gauge-ts` we will simply rename the file to `.ts` and refactor that file.
> 
> here is the sample ts file.
> 
> image

If this is a ts file, then why isn't it throwing error for undefined name step, beforeScenario etc and for not providing a return type to async function.

**Comment by h2grover** [comment]
@anjaliahlawat please create `global.d.ts` file and add this. Also please post your thought on what is `global.d.ts` file?

```
declare const beforeScenario: any;
declare const step: any;
declare const gauge: any;
declare const document: any;
declare const beforeSuite: any;
declare const beforeSpec: any;
declare const afterScenario: any;
declare const afterSpec: any;
declare const afterSuite: any;
declare const afterStep: any;
```

**Comment by anjaliahlawat** [comment]
> @anjaliahlawat please create `global.d.ts` file and add this. Also please post your thought on what is `global.d.ts` file?
> 
> ```
> declare const beforeScenario: any;
> declare const step: any;
> declare const gauge: any;
> declare const document: any;
> declare const beforeSuite: any;
> declare const beforeSpec: any;
> declare const afterScenario: any;
> declare const afterSpec: any;
> declare const afterSuite: any;
> declare const afterStep: any;
> ```

I supposed it's a declaration file where we are declaring these functions globally. 

Also, If we are using any, aren't violating the core principle of typescript? 
Plus eslint shows warning on using any. Or should I disable this warning for this particular file?

**Comment by h2grover** [comment]
@anjaliahlawat  since we are not using `gauge-ts` and we can't really define the step types as it is managed by `gauge` itself. We have to focus on converting the code written by us. Once `gauge-ts` is public we can think about it at that time.

And vscode must be giving a warning for `any` type not an error. so you can ignore that for now

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "target": "ES2019",
+        "moduleResolution": "node",
+        "outDir": "./build",
```
Comment (h2grover) [inline tsconfig.json:6]: @anjaliahlawat Please try to run the code now and after that please add this line and resolve the errors.
 `"strict": true,`

**Review by h2grover — CHANGES_REQUESTED** [review]

**Comment by h2grover** [comment]
@anjaliahlawat are you done with the changes?

**Comment by anjaliahlawat** [comment]
> @anjaliahlawat are you done with the changes?

Yes.

**Comment by h2grover** [comment]
@anjaliahlawat please make sure you click this button or ask me to review by mentioning me on the PR
image

File: tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+        "strict": true,
+    },
+    "include": [
+        "./src/**/*",
+    ],
+}
```
Comment (h2grover) [inline tsconfig.json]: also, please self-review the PR first. I don't want to point you again and again.

Please add new line

File: src/tests/Setup.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import * as puppeteer from "puppeteer";
```
Comment (h2grover) [inline src/tests/Setup.ts]: please import specific object

**Comment by h2grover** [comment]
@anjaliahlawat can you please post the job URL here after testing this branch

**Comment by anjaliahlawat** [comment]
> @anjaliahlawat can you please post the job URL here after testing this branch

@h2grover https://github.com/anjaliahlawat/amazon-login-script/runs/2246167846?check_suite_focus=true

**Comment by h2grover** [comment]
@anjaliahlawat thanks, please remind me, haven't you fixed the failed test case earlier ?

**Comment by anjaliahlawat** [comment]
> @anjaliahlawat thanks, please remind me, haven't you fixed the failed test case earlier ?

No I couldn't resolve it. So you told me to move on and work on further tasks.

**Review by h2grover — APPROVED** [review]

**Pull request merged**